lib/apiservers/engine/backends/convert/stats.go

```diff
@@ -0,0 +1,266 @@
+				close(metric)
+				cs.Stop()
+				return
+			default:
+				for dec.More() {
+					err := dec.Decode(&vmm)
+					if err != nil {
+						log.Errorf("container metric decoding error for container(%s): %s", cs.config.ContainerID, err)
+						cs.config.Cancel()
+					}
+					// send the decoded metric for transform and encoding
+					metric <- vmm
+				}
+			}
+		}
+
+	}()
+
+	go func() {
+		for {
+			select {
+			case <-cs.config.Ctx.Done():
+				return
+			case <-finished:
+				return
+			case nm := <-metric:
+				// convert the Stat to docker struct
+				stats, err := cs.ToContainerStats(&nm)
+				if err != nil {
+					log.Errorf("container metric conversion error for container(%s): %s", cs.config.ContainerID, err)
+					cs.config.Cancel()
+				}
+				// Do we have a complete stat that can be sent to the client?
+				if stats != nil {
+					err = doc.Encode(stats)
+					if err != nil {
+						log.Warnf("container metric encoding error for container(%s): %s", cs.config.ContainerID, err)
+						cs.config.Cancel()
+					}
+					// if we aren't streaming then cancel
+					if !cs.config.Stream {
+						cs.config.Cancel()
+					}
+					// set to previous stat so we can reuse
+					previousStat = stats
+				}
+			default:
+				// the docker client expects updates quicker than vSphere can produce them, so
+				// we need to send the previous stats to avoid intermittent empty output
+				time.Sleep(time.Second * 1)
+				if previousStat != nil && cs.reader != nil {
+					err := doc.Encode(previousStat)
+					if err != nil {
+						log.Warnf("container previous metric encoding error for container(%s): %s", cs.config.ContainerID, err)
+						cs.config.Cancel()
+					}
+				}
+			}
+		}
+	}()
+	return cs.writer
+}
+
+// ToContainerStats will convert the vic VMMetrics to a docker stats struct -- a complete docker stats
+// struct requires two samples.  Func will return nil until a complete stat is available
+func (cs *ContainerStats) ToContainerStats(current *metrics.VMMetrics) (*types.StatsJSON, error) {
+	// if we have a current metric then validate and transform
+	if cs.currentMetrics != nil {
+		// do we have the same metric as before?
+		if cs.currentMetrics.SampleTime.Equal(current.SampleTime) {
+			// we've already got this as current, so skip and wait for the
+			// next sample
+			return nil, nil
+		}
+		// we have new current stats so need to move the previous CPU
+		err := cs.previousCPU(current)
+		if err != nil {
+			return nil, err
+		}
+	}
+	cs.currentMetrics = current
+
+	// create the current CPU stats
+	cs.currentCPU()
+
+	// create memory stats
+	cs.memory()
+
+	// set sample time
+	cs.dockerStats.Read = cs.currentMetrics.SampleTime
+
+	// PreRead will be zero if we don't have two samples
+	if cs.dockerStats.PreRead.IsZero() {
+		return nil, nil
+	}
+	return cs.dockerStats, nil
+}
+
+func (cs *ContainerStats) memory() {
+	// given KB (i.e. 2048) convert to GB
+	cs.dockerStats.MemoryStats.Limit = uint64(cs.config.Memory * 1000 * 1000)
```

Comment:
For memory limit we are given MB not KB -- comment changed

For active memory we are given KB, so multiplier changed to 1024